datasource/proxmox/data.go

```diff
@@ -0,0 +1,412 @@
+			if err != nil {
+				return nil, err
+			}
+			if vmCtime > maxCtime {
+				maxCtime = vmCtime
+				result = configs[i]
+			}
+		} else {
+			return nil, errors.New("no meta field in the guest config")
+		}
+	}
+	return result, nil
+}
+
+// Get configs from PVE in 'map[string]interface{}' format for all VMs in the list.
+// Also add value of VM ID to every config (useful for further steps).
+func getVmConfigs(client *proxmox.Client, vmList []proxmox.GuestResource) ([]vmConfig, error) {
+	var result []vmConfig
+	for _, vm := range vmList {
+		var thisConfig vmConfig
+		vmr := proxmox.NewVmRef(int(vm.Id))
+		thisConfig, err := client.GetVmConfig(vmr)
+		if err != nil {
+			return nil, err
+		}
+		thisConfig["vmid"] = vm.Id
+		result = append(result, thisConfig)
+	}
+	return result, nil
+}
+
+// Drop guests from list that are not match some filters in the datasource config.
+func filterGuests(config Config, guests []proxmox.GuestResource) []proxmox.GuestResource {
+	var result []proxmox.GuestResource
+
+	if config.Name != "" {
+		result = filterByName(guests, config.Name)
+	} else {
+		result = guests
+	}
+
+	if config.NameRegex != "" {
+		result = filterByNameRegex(guests, config.NameRegex)
+	} else {
+		if config.Name == "" {
+			result = guests
+		}
+	}
+
+	if config.Template {
+		result = filterByTemplate(result)
+	}
+	if config.Node != "" {
+		result = filterByNode(result, config.Node)
+	}
+	if config.VmTags != "" {
+		result = filterByTags(result, config.VmTags)
+	}
+
+	return result
+}
+
+func filterByName(guests []proxmox.GuestResource, name string) []proxmox.GuestResource {
+	result := make([]proxmox.GuestResource, 0)
+	for _, i := range guests {
+		if i.Name == name {
+			result = append(result, i)
+		}
+	}
+	return result
+}
+
+func filterByNameRegex(guests []proxmox.GuestResource, nameRegex string) []proxmox.GuestResource {
+	re, _ := regexp.Compile(nameRegex)
+	result := make([]proxmox.GuestResource, 0)
+	for _, i := range guests {
+		if re.MatchString(i.Name) {
+			result = append(result, i)
+		}
+	}
+	return result
+}
+
+func filterByTemplate(guests []proxmox.GuestResource) []proxmox.GuestResource {
+	result := make([]proxmox.GuestResource, 0)
+	for _, i := range guests {
+		if i.Template {
+			result = append(result, i)
+		}
+	}
+	return result
+}
+
+func filterByNode(guests []proxmox.GuestResource, node string) []proxmox.GuestResource {
+	result := make([]proxmox.GuestResource, 0)
+	for _, i := range guests {
+		if i.Node == node {
+			result = append(result, i)
+		}
+	}
+	return result
+}
+
+func filterByTags(guests []proxmox.GuestResource, tags string) []proxmox.GuestResource {
+	result := make([]proxmox.GuestResource, 0)
+	// Split tags string because it can contain several tags separated with ";"
+	tagsSplitted := strings.Split(tags, ";")
+	for _, guest := range guests {
+		if len(guest.Tags) > 0 && configTagsMatchNodeTags(tagsSplitted, guest.Tags) {
+			result = append(result, guest)
+		}
+	}
+	return result
+}
+
```

Comment:
Hi @mogrogan! Thank you for the review!
Your suggestion seems reasonable. I will rewrite the `filterGuests` function.

Also, after a while and after implementing a datasource for another Packer plugin, I want to rename this datasource because in the future it may not be the only datasource for the Proxmox plugin. And this name may confuse other people.